src/shared/sleep-config.c

```diff
@@ -178,56 +180,79 @@ int find_hibernate_location(char **device, char **type, size_t *size, size_t *us
+                if (i > MAX_SWAP_ENTRIES) {
+                        log_debug("More than %i entries found in /proc/swaps; ignoring remaining", MAX_SWAP_ENTRIES);
+                        break;
+                }
+
+                SwapEntry *swap;
                 int k;
 
+                swap = new0(SwapEntry, 1);
+
                 k = fscanf(f,
                            "%ms "   /* device/file */
                            "%ms "   /* type of swap */
                            "%zu "   /* swap size */
                            "%zu "   /* used */
-                           "%*i\n", /* priority */
-                           &dev_field, &type_field, &size_field, &used_field);
+                           "%i\n",  /* priority */
+                           &swap->device, &swap->type, &swap->size, &swap->used, &swap->priority);
                 if (k == EOF)
                         break;
-                if (k != 4) {
+                if (k != 5) {
                         log_warning("Failed to parse /proc/swaps:%u", i);
                         continue;
                 }
 
-                if (streq(type_field, "file")) {
+                if (streq(swap->type, "file")) {
 
-                        if (endswith(dev_field, "\\040(deleted)")) {
-                                log_warning("Ignoring deleted swap file '%s'.", dev_field);
+                        if (endswith(swap->device, "\\040(deleted)")) {
+                                log_warning("Ignoring deleted swap file '%s'.", swap->device);
                                 continue;
                         }
 
-                } else if (streq(type_field, "partition")) {
+                } else if (streq(swap->type, "partition")) {
                         const char *fn;
 
-                        fn = path_startswith(dev_field, "/dev/");
+                        fn = path_startswith(swap->device, "/dev/");
                         if (fn && startswith(fn, "zram")) {
-                                log_debug("Ignoring compressed RAM swap device '%s'.", dev_field);
+                                log_debug("Ignoring compressed RAM swap device '%s'.", swap->device);
                                 continue;
                         }
                 }
 
-                if (device)
-                        *device = TAKE_PTR(dev_field);
-                if (type)
-                        *type = TAKE_PTR(type_field);
-                if (size)
-                        *size = size_field;
-                if (used)
-                        *used = used_field;
-                return 0;
+                /* insert swap into array in priority order */
+                s = swap_count;
+                while ((s > 0) && (swap->priority > swaps[swap_count-1]->priority)) {
+                        swaps[s] = swaps[s-1];
+                        s--;
+                }
+
+                swaps[s] = swap;
+                swap_count++;
```

Comment:
why do you keep the array at all? unless I am missing something you never care about entries > 1, you only care about entries 0. Hence you might as well never store those, but simple parse an entry, check if it has a higher priority than the previously stored one, and if so, store it, and otherwise forget it, and the proceed to the next